### PR TITLE
Support up-to-date `botocore`

### LINKED ISSUE #1065: Support botocore>=1.33.2
Creating an issue to track the implementation mentioned in https://github.com/aio-libs/aiobotocore/pull/1057#issuecomment-1830899886.

In particular, we are eager to try out S3 Express One Zone!

### PR DESCRIPTION
**Pull request by jakob-keller**
### Description of Change
This PR intends to improve general compatibility of `aiobotocore` within the Python ecosystem by bumping the dependency specification of `botocore`, as well as `boto3` and `awscli`.

It also restores feature parity with respect to recently released and improved AWS services, in particular adding support for [S3 Express One Zone](https://aws.amazon.com/blogs/aws/new-amazon-s3-express-one-zone-high-performance-storage-class/).

### Assumptions
[Upstream diff](https://github.com/boto/botocore/compare/1.33.1..1.33.13) contains multiple changes that affect this codebase.

### Checklist for All Submissions
* [x] I have added change info to [CHANGES.rst](https://github.com/aio-libs/aiobotocore/blob/master/CHANGES.rst)
* [x] If this is resolving an issue (needed so future developers can determine if change is still necessary and under what conditions) (can be provided via link to issue with these details): closes #1065 
  * [ ] Detailed description of issue
  * [ ] Alternative methods considered (if any)
  * [ ] How issue is being resolved
  * [ ] How issue can be reproduced
* [ ] If this is providing a new feature  (can be provided via link to issue with these details):
  * [ ] Detailed description of new feature
  * [ ] Why needed
  * [ ] Alternatives methods considered (if any)

### Checklist when updating botocore and/or aiohttp versions

* [x] I have read and followed [CONTRIBUTING.rst](https://github.com/aio-libs/aiobotocore/blob/master/CONTRIBUTING.rst#how-to-upgrade-botocore)
* [x] I have updated test_patches.py where/if appropriate (also check if no changes necessary)
* [x] I have ensured that the awscli/boto3 versions match the updated botocore version

### CODE REVIEW

**Comment by jakob-keller** [comment]
@thehesiod: Another bump that adds support for new AWS services and features.

I'm unsure how to deal with the lack in test coverage. Is this something to be concerned about?

**Comment by thehesiod** [comment]
> @thehesiod: Another bump that adds support for new AWS services and features.
> 
> I'm unsure how to deal with the lack of test coverage. Is this something to be concerned about?

not reaaaly, bonus if they're added.  We should probably have a task to port all the botocore tests over

**Comment by thehesiod** [comment]
will get to this today

**Comment by jakob-keller** [comment]
> will get to this today

I just pushed another update that extends compatibility up to yesterday's botocore release. It's all yours now ;-)

**Comment by thehesiod** [comment]
whoops didn't get to it. will do asap

File: setup.py [hunk 1]
```diff
@@ -7,15 +7,15 @@
 # NOTE: When updating botocore make sure to update awscli/boto3 versions below
 install_requires = [
     # pegged to also match items in `extras_require`
-    'botocore>=1.32.4,<1.33.2',
```
Comment (thehesiod) [inline setup.py:10]: I missed this in the previous review, this was dangerous as there were changes to aiobotocore needed to match the latest botocore, so I'm not sure they were all backwards compatible

File: setup.py [hunk 2]
```diff
@@ -7,15 +7,15 @@
...
+    'botocore>=1.33.2,<1.33.14',
```
Comment (thehesiod) [inline setup.py:10]: given the various internal changes required, I think this should be `botocore>=1.33.13,<1.33.14` unless you worked through all the backwards compatibility impacts

Comment (thehesiod) [reply setup.py:10]: also note we only test one version of botocore

Comment (jakob-keller) [reply setup.py:10]: The PR should be consistent in that regard. Upstream code is stable for that version range.

Comment (thehesiod) [reply setup.py:10]: ah, I see what you did, the issues is from 1.33.1->1.33.2 nice 👍🏼

File: aiobotocore/utils.py [hunk 3]
```diff
@@ -348,6 +352,72 @@
...
+                    next(iter(self._aio_credential_cache)),
+                )
+
+            self._aio_credential_cache[bucket] = credentials
+
+        return credentials
+
+    def build_refresh_callback(self, bucket):
+        async def refresher():
+            response = await self._client.create_session(Bucket=bucket)
```
Comment (thehesiod) [inline aiobotocore/utils.py:397]: I can't find this method, is this actually async?

Comment (thehesiod) [reply aiobotocore/utils.py:397]: ah, must be constructed from the json somehow

**Comment by thehesiod** [comment]
going to do version w/o prefix this time, see how it goes

File: aiobotocore/utils.py (hunk 3 above)
Comment (jakob-keller) [reply aiobotocore/utils.py:397]: Correct, I struggled with that, too. It’s a new S3 API.

**Comment by thehesiod** [comment]
thanks for all the help!

File: setup.py (hunk 1 above)
Comment (jakob-keller) [reply setup.py:10]: There were no backwards incompatible changes in that previous version range.

**Comment by thehesiod** [comment]
I liked your strategy of supporting multiple botocore versions, probably worth adding to the docs describing how to do upgrades

**Comment by thehesiod** [comment]
actually was thinking about this last night, we can probably add a python helper where you set a base revision, and it tells you which ranges have only service json changes.  Would save a lot of manual labor.

**Comment by jakob-keller** [comment]
> we can probably add a python helper where you set a base revision, and it tells you which ranges have only service json changes.

That's true. To reach the next level of automation, it would be nice to run a scheduled GitHub action that checks for new botocore releases and opens PRs, if these are light-weight, i.e. only contain service JSON updates. Once that is stable, these PRs could even be auto-merged after passing CI. Eventually, we could also do automated minor releases.

**Comment by thehesiod** [comment]
i like your thinking :]

**Pull request merged**